PyExpLabSys/common/value_logger.py

```diff
@@ -1,10 +1,285 @@
-""" Reads continuously updated values and decides whether it is time to log a new point """
+""" Reads continuously updated values and decides whether it is time to log a new point 
 
+Two loggers (or checkers) exist: ValueLogger and LoggingCriteriumChecker. The internal
+workings are slightly different, but both checks the newest measured value against the
+last recorded value and determines whether or not the newest value should be saved.
+    A timeout counter (default 600s) forces data to be saved at no longer than this
+interval.
+    An event checker (logarithmic or linear) checks for events based on a large enough
+change since the last recorded value.
+The amount of data saved using these, can be down to 0.1% to 3% of the full dataset
+while retaining key features (details around events).
+
+ *LoggingCriteriumChecker* is an object class capable of keeping track of saved values
+for several different datasets - each represented by a unique codename. Its checks are
+manually handled for every data point in any thread handling the collected data. This
+checker is responsible for timestamping the data value.
+
+All data points between events are buffered in the instance to be able to retrospective-
+ly save relevant data. On an event, the code loops through the buffer from the newest
+measurement to follow the slope down to the beginning of the event and saves every data
+point based on a subcriterium (default is 10% of the normal criterium). This takes care
+of most details around the event. For even finer details, the remaining buffer is run
+through the check_buffer_deviation method, which checks whether a linear representation
+of the data deviates significantly from the average linear representation of data
+gathered from timeout events. Any significantly deviant data is added to the save pool.
+These data points can be served the timestamp which also makes it possible to use this
+module as a filter on existing data.
+
+Example 1: Use the LoggingCriteriumChecker to filter pressure and speed before logging
+--------------------
+from PyExpLabSys.common.value_logger import LoggingCriteriumChecker
+
+# Create an instance
+logger = LoggingCriteriumChecker(
+    codenames=['pressure', 'speed'],
+    types=['log', 'lin'],
+    criteria=[0.25, 5],
+    time_outs=[600, 600],
+    grades=[0.1, 0.1],
+)
+
+# Pressure events are detected at 25% value change and timeouts of 600s. 2.5% is used as
+# the subcriterium. Speed events are detected at 5 unit changes and timeouts of 600s.
+# 0.5 units are used as the subcriterium.
+
+# Use in some code
+while True:
+    for codename in ['pressure', 'speed']:
+
+        # Get newest timestamped data either directly from the driver unit or from an
+        # existing data set
+        time_, value = get_newest_data(codename)
```

Comment:
I am not familiar with the syntax of using an underscore as postfix? If this is just to avoid a name clash with the `time` module, I will suggest to use different solution, such as naming the variable `timestamp` or `datetime` or similar.

Reply:
Good idea. Yes it is just because I prefer to use the short variable name, but it clashes with the module. I guess I can get used to the slightly longer "timestamp" ;)